backends-common/postgres/src/test/java/org/apache/james/backends/postgres/SimpleJamesPostgresConnectionFactoryTest.java

```diff
@@ -0,0 +1,142 @@
+            .host(container.getHost())
+            .port(container.getMappedPort(5432))
+            .username(DB_USER)
+            .password(DB_PASSWORD)
+            .database(DB_NAME)
+            .build());
+        jamesPostgresConnectionFactory = new SimpleJamesPostgresConnectionFactory(connectionFactory);
+    }
+
+    @Test
+    void factoryShouldCreateCorrectNumberOfConnections() {
+        PostgresqlConnection connection = connectionFactory.create().block();
+        Integer previousDbActiveNumberOfConnections = getNumberOfConnections(connection);
+
+        // create 50 connections
+        Flux.range(1, 50)
+            .flatMap(i -> jamesPostgresConnectionFactory.getConnection(Domain.of("james"+i)))
+            .last()
+            .block();
+
+        Integer dbActiveNumberOfConnections = getNumberOfConnections(connection);
+
+        assertThat(dbActiveNumberOfConnections - previousDbActiveNumberOfConnections).isEqualTo(50);
+    }
+
+    @Nullable
+    private static Integer getNumberOfConnections(PostgresqlConnection connection) {
+        return Mono.from(connection.createStatement("SELECT count(*) from pg_stat_activity where usename = $1;")
+            .bind("$1", DB_USER)
+            .execute()).flatMap(result -> Mono.from(result.map((row, rowMetadata) -> row.get(0, Integer.class)))).block();
+    }
+
+    @Test
+    void factoryShouldNotCreateNewConnectionWhenDomainsAreTheSame() {
+        Domain domain = Domain.of("james");
+        Connection connectionOne = jamesPostgresConnectionFactory.getConnection(domain).block();
+        Connection connectionTwo = jamesPostgresConnectionFactory.getConnection(domain).block();
+
+        assertThat(connectionOne == connectionTwo).isTrue();
+    }
+
+    @Test
+    void factoryShouldNotCreateNewConnectionWhenDomainsAreTheSameAndRequestsAreFromDifferentThreads() throws Exception {
+        Set<Connection> connectionSet = ConcurrentHashMap.newKeySet();
+
+        ConcurrentTestRunner.builder()
+            .reactorOperation((threadNumber, step) -> jamesPostgresConnectionFactory.getConnection(Domain.of("james"))
+                .doOnNext(connectionSet::add)
+                .then())
+            .threadCount(100)
+            .operationCount(10)
+            .runSuccessfullyWithin(Duration.ofMinutes(1));
+
+        assertThat(connectionSet).hasSize(1);
+    }
+
+    @Test
+    void factoryShouldCreateOnlyOneDefaultConnection() throws Exception {
+        Set<PostgresqlConnection> connectionSet = ConcurrentHashMap.newKeySet();
+
+        ConcurrentTestRunner.builder()
+            .reactorOperation((threadNumber, step) -> jamesPostgresConnectionFactory.getConnection(Optional.empty())
+                .doOnNext(connectionSet::add)
+                .then())
+            .threadCount(100)
+            .operationCount(10)
+            .runSuccessfullyWithin(Duration.ofMinutes(1));
+
+        assertThat(connectionSet).hasSize(1);
+    }
+
+}
```

Comment:
I think we need test for:

1. 
Given domain1, domain2
When get connection domain1, domain2
then Connection1 isNotEqual Connection2 


2. 
Given domain1, domain2
When get connection domain1, domain2
Then SQL `show app.current_domain` of connection1 isNotEqual `SHOW app.current_domain` of connection 2 

3. 
Check result of `SHOW app.current_domain` for `Default connection`

Reply:
added